File: server/app/controllers/questions.py
import logging

from fastapi import APIRouter, Depends
from supabase._async.client import AsyncClient as Client

from app.models.question import (
    AssociateQuestionWithChallengeRequest,
    CreateTemplateQuestionRequest,
    Question,
    RunTestsRequest,
)
from app.services.questions import QuestionsService
from app.utils.dependencies import init_db_client

log = logging.getLogger(__name__)


class QuestionsController:
    def __init__(self, service: QuestionsService):
        self.router = APIRouter()
        self.service = service
        self.setup_routes()

    def setup_routes(self):
        router = self.router

        @router.get(
            "",
            response_model=list[Question],
        )
        async def get_all_questions(client: Client = Depends(init_db_client)) -> list[Question]:
            log.info("Getting all questions...")
            response: list[Question] = await self.service.get_all_questions(client=client)
            log.info("Questions: %s", response)
            return response

        @router.post(
            "/template",
            response_model=Question,
        )
        async def create_template_question(
            input: CreateTemplateQuestionRequest, client: Client = Depends(init_db_client)
        ) -> Question:
            log.info("Creating template question...")
            response: Question = await self.service.create_template_question(
                input=input, client=client
            )
            log.info("Template question: %s", response)
            return response

        @router.get(
            "/question_id/{question_id}",
            response_model=Question,
        )
        async def get_question_by_id(
            question_id: str, client: Client = Depends(init_db_client)
        ) -> Question:
            log.info(f"Getting question with id {question_id}...")
            response: Question = await self.service.get_question_by_id(
                question_id=question_id, client=client
            )
            log.info("Question: %s", response)
            return response

        @router.get(
            "/challenge_id/{challenge_id}",
            response_model=list[Question],
        )
        async def get_questions_by_challenge_id(
            challenge_id: str, client: Client = Depends(init_db_client)
        ) -> list[Question]:
            log.info(f"Getting all questions associated with challenge id {challenge_id}...")
            response: list[Question] = await self.service.get_questions_by_challenge_id(
                challenge_id=challenge_id, client=client
            )
            log.info("Questions: %s", response)
            return response

        @router.post(
            "/associate",
        )
        async def associate_question_with_challenge(
            input: AssociateQuestionWithChallengeRequest,
            client: Client = Depends(init_db_client),
        ) -> None:
            log.info("Associating question with challenge...")
            await self.service.associate_question_with_challenge(input=input, client=client)
            log.info("Question associated with challenge successfully")

        @router.post(
            "/run-tests/{question_id}",
        )
        async def run_tests(
            question_id: str,
            run_tests_request: RunTestsRequest,
            client: Client = Depends(init_db_client),
        ) -> list[str]:
            log.info(f"Running tests for question {question_id}...")
            result: list[str] = await self.service.run_tests(
                question_id=question_id, code=run_tests_request.code, client=client
            )
            log.info("Tests run successfully")
            return result

File: forge/src/actions/questions.ts
'use server';

import axios from 'axios';

import {
  AssociateQuestionWithChallengeRequest,
  CreateTemplateQuestionRequest,
  Question,
  question,
} from '@/types/questions';

import { getClerkToken } from '@/lib/clerk';

export async function createTemplateQuestion(
  input: CreateTemplateQuestionRequest,
): Promise<Question> {
  const token: string = await getClerkToken();

  try {
    console.log('Creating template question...');
    const response = await axios.post(
      `${process.env.EXPO_PUBLIC_API_BASE_URL}/api/questions/template`,
      input,
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
      },
    );
    console.log('Successfully created template question');
    return question.parse(response.data);
  } catch (error) {
    console.error('Error creating template question:', error);
    throw error;
  }
}

export async function getQuestionById(questionId: string): Promise<Question> {
  const token: string = await getClerkToken();

  try {
    console.log('Getting question...');
    const response = await axios.get(
      `${process.env.EXPO_PUBLIC_API_BASE_URL}/api/questions/question_id/${questionId}`,
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
      },
    );
    console.log('Get Question By Id Response', response.data);
    return question.parse(response.data);
  } catch (error) {
    console.error('Error getting question by id:', error);
    throw error;
  }
}

export async function getQuestionsByChallengeId(challengeId: string): Promise<Question[]> {
  const token: string = await getClerkToken();

  try {
    console.log('Getting questions associated with challenge id...');
    const response = await axios.get(
      `${process.env.EXPO_PUBLIC_API_BASE_URL}/api/questions/challenge_id/${challengeId}`,
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
      },
    );
    console.log('Get Questions By Challenge Id Response', response.data);
    return response.data.map((q: Question) => question.parse(q));
  } catch (error) {
    console.error('Error getting questions by challenge id:', error);
    throw error;
  }
}

export async function getAllQuestions(): Promise<Question[]> {
  const token: string = await getClerkToken();

  try {
    console.log('Getting all questions...');
    const response = await axios.get(`${process.env.EXPO_PUBLIC_API_BASE_URL}/api/questions`, {
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
    });
    console.log('Successfully retrieved all questions');
    return response.data.map((q: Question) => question.parse(q));
  } catch (error) {
    console.error('Error getting all questions:', error);
    throw error;
  }
}

export async function associateQuestionWithChallenge(
  input: AssociateQuestionWithChallengeRequest,
): Promise<void> {
  const token: string = await getClerkToken();

  try {
    console.log('Associating question with challenge...');
    await axios.post(`${process.env.EXPO_PUBLIC_API_BASE_URL}/api/questions/associate`, input, {
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
    });
    console.log('Successfully associated question with challenge');
  } catch (error) {
    console.error('Error associating question with challenge:', error);
    throw error;
  }
}
